upup/pkg/fi/cloudup/hetzner/cloud.go

```diff
@@ -257,21 +301,118 @@ func (c *hetznerCloudImplementation) Region() string {
 	return c.region
 }
 
-func (c *hetznerCloudImplementation) GetCloudGroups(cluster *kops.Cluster, instancegroups []*kops.InstanceGroup, warnUnmatched bool, nodes []v1.Node) (map[string]*cloudinstances.CloudInstanceGroup, error) {
-	// TODO(hakman): Implement me
-	return nil, fmt.Errorf("hetzner cloud provider does not implement GetCloudGroups at this time")
+func (c *hetznerCloudImplementation) GetCloudGroups(cluster *kops.Cluster, instanceGroups []*kops.InstanceGroup, warnUnmatched bool, nodes []v1.Node) (map[string]*cloudinstances.CloudInstanceGroup, error) {
+	nodeMap := cloudinstances.GetNodeMap(nodes, cluster)
+
+	serverGroups, err := findServerGroups(c, cluster.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find server groups: %v", err)
+	}
+
+	cloudInstanceGroups := make(map[string]*cloudinstances.CloudInstanceGroup)
+	for name, serverGroup := range serverGroups {
+		var instanceGroup *kops.InstanceGroup
+		for _, ig := range instanceGroups {
+			groupName := fmt.Sprintf("%s-%s", cluster.Name, ig.Name)
+			if name == groupName {
+				instanceGroup = ig
+				break
+			}
+		}
+		if instanceGroup == nil {
+			if warnUnmatched {
+				klog.Warningf("Server group %q has no corresponding instance group", name)
+			}
+			continue
+		}
+
+		cloudInstanceGroups[instanceGroup.Name], err = buildCloudInstanceGroup(instanceGroup, serverGroup, nodeMap)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build cloud instance group for instance group %q: %w", instanceGroup.Name, err)
+		}
+	}
+
+	return cloudInstanceGroups, nil
+}
+
+// findServerGroups finds all server groups belonging to the cluster
+func findServerGroups(c *hetznerCloudImplementation, clusterName string) (map[string][]*hcloud.Server, error) {
```

Comment:
Nit: if you're only calling this from GetCloudGroups, I think you could pass the instancegroups in and directly match the instancegroups, instead of constructing the magic string (?)